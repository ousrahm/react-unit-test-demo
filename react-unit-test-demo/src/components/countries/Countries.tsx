import React from "react";
import styles from "./Countries.module.css";
import { Button, Col, Container, Form, Row } from "react-bootstrap";

export interface CountryName {
  name: {
    common: string;
  };
}

const Countries = (): JSX.Element => {
  const [names, setNames] = React.useState<string[]>([]);
  const [language, setLanguage] = React.useState("");
  const [error, setError] = React.useState("");

  const handleSearch = React.useCallback(async (lang: string) => {
    try {
      await fetch(
        `https://restcountries.com/v3.1/lang/${lang.trim()}?fields=name`
      ).then(async (response) => {
        setError("");
        const countries: CountryName[] = await response.json();

        const countryNames: string[] = countries.map((c) => c.name.common);

        setNames(countryNames);
      });
    } catch {
      setNames([]);
      setError(
        "Uh oh! There was an issue retrieving your data. Make sure you spelled it correctly."
      );
    }
  }, []);

  return (
    <div className="Countries">
      <h2 className={styles.heading}>Countries By Language</h2>
      <h4 className={styles.instructions}>
        Enter the name of a language and press Search to retrieve a list of
        countries with that as their official language.
      </h4>
      <Container>
        <Row>
          <Col>
            <ul className={styles.list}>
              {names.map((n) => (
                <li key={n}>{n}</li>
              ))}
            </ul>
          </Col>
          <Col>
            <Row className={styles.languageSearchRow}>
              <Form.Control
                className={styles.languageInput}
                type="text"
                onChange={(e) => setLanguage(e.target.value)}
                placeholder="language here!"
              />
              <Button
                className={styles.button}
                onClick={() => handleSearch(language)}
              >
                Search
              </Button>
            </Row>
            <Row className={styles.error}>
              <span>{error}</span>
            </Row>
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default Countries;
