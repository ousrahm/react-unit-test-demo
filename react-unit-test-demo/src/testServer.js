import { setupServer } from 'msw/node'
// This configures a request mocking server with the given request handlers.
export const server = setupServer(...[]);

// eslint-disable-next-line import/no-anonymous-default-export
export default {};