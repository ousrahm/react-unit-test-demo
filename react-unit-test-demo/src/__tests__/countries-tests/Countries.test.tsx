import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Countries, { CountryName } from "../../components/countries/Countries";
import { server } from "../../testServer";
import { rest } from "msw";

export const countryNames: CountryName[] = [
  { name: { common: "Italy" } },
  { name: { common: "San Marino" } },
  { name: { common: "Switzerland" } },
  { name: { common: "Vatican City" } },
];

describe("Countries component tests", () => {
  beforeAll(() => server.listen());
  //   beforeEach(() => jest.spyOn(global, "fetch").mockImplementation(jest.fn()));
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it("renders without crashing", () => {
    render(<Countries />);
  });

  it("matches snapshot", () => {
    const countries = render(<Countries />).baseElement;

    expect(countries).toMatchSnapshot();
  });

  it("displays italy when called w/ italian", async () => {
    server.use(
      rest.get("https://restcountries.com/v3.1/lang/italian", (req, res, ctx) =>
        res(ctx.status(200), ctx.json(countryNames))
      ),
      rest.get("https://restcountries.com/v3.1/lang/klingon", (req, res, ctx) =>
        res(ctx.status(400), ctx.json(countryNames))
      )
    );

    render(<Countries />);

    const languageInput = await screen.findByPlaceholderText("language here!");
    const searchButton = await screen.findByText("Search");

    fireEvent.change(languageInput, { target: { value: "italian" } });
    fireEvent.click(searchButton);

    const italy = await screen.findByText("Italy");
    const sanMarino = await screen.findByText("San Marino");
    const switzerland = await screen.findByText("Switzerland");
    const vaticanCity = await screen.findByText("Vatican City");

    expect(italy).toBeInTheDocument();
    expect(sanMarino).toBeInTheDocument();
    expect(switzerland).toBeInTheDocument();
    expect(vaticanCity).toBeInTheDocument();
  });

  it("displays the error message when called w/ klingon", async () => {
    server.use(
      rest.get("https://restcountries.com/v3.1/lang/klingon", (req, res, ctx) =>
        res(ctx.status(400), ctx.json("ERROR!"))
      )
    );

    render(<Countries />);

    const languageInput = await screen.findByPlaceholderText("language here!");
    const searchButton = await screen.findByText("Search");

    fireEvent.change(languageInput, { target: { value: "klingon" } });
    fireEvent.click(searchButton);

    const errorMessage = await screen.findByText(
      "Uh oh! There was an issue retrieving your data. Make sure you spelled it correctly."
    );

    expect(errorMessage).toBeInTheDocument();
  });
});
