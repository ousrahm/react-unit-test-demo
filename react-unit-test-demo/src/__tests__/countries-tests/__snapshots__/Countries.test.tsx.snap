// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries component tests matches snapshot 1`] = `
<body>
  <div>
    <div
      class="Countries"
    >
      <h2
        class="heading"
      >
        Countries By Language
      </h2>
      <h4
        class="instructions"
      >
        Enter the name of a language and press Search to retrieve a list of countries with that as their official language.
      </h4>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <ul
              class="list"
            />
          </div>
          <div
            class="col"
          >
            <div
              class="languageSearchRow row"
            >
              <input
                class="languageInput form-control"
                placeholder="language here!"
                type="text"
              />
              <button
                class="button btn btn-primary"
                type="button"
              >
                Search
              </button>
            </div>
            <div
              class="error row"
            >
              <span />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
